collector/btrfs_linux.go

```diff
@@ -0,0 +1,187 @@
+}
+
+// updateBtrfsStats collects statistics for one bcache ID.
+func (c *btrfsCollector) updateBtrfsStats(ch chan<- prometheus.Metric, s *btrfs.Stats) {
+	const subsystem = "btrfs"
+
+	// Basic information about the filesystem
+	devLabels := []string{"label", "uuid"}
+
+	// Retrieve the metrics
+	metrics := c.getMetrics(s)
+
+	// Convert all gathered metrics to Prometheus Metrics and add to channel.
+	for _, m := range metrics {
+		labels := append(devLabels, m.extraLabel...)
+
+		desc := prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, m.name),
+			m.desc,
+			labels,
+			nil,
+		)
+
+		labelValues := []string{s.Label, s.UUID}
+		if len(m.extraLabelValue) > 0 {
+			labelValues = append(labelValues, m.extraLabelValue...)
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			desc,
+			prometheus.GaugeValue,
+			m.value,
+			labelValues...,
+		)
+	}
+}
+
+// getMetrics returns metrics for the given Btrfs statistics
+func (c *btrfsCollector) getMetrics(s *btrfs.Stats) []btrfsMetric {
+	metrics := []btrfsMetric{
+		{
+			name:  "devices",
+			desc:  "Number of devices that are part of the filesystem.",
+			value: float64(len(s.Devices)),
+		},
+		{
+			name:  "global_rsv_size_bytes",
+			desc:  "Size of global reserve.",
+			value: float64(s.Allocation.GlobalRsvSize),
+		},
+	}
+
+	// Information about devices
+	for n, dev := range s.Devices {
+		metrics = append(metrics, btrfsMetric{
+			name:            "device_size",
+			desc:            "Size of a device that is part of the filesystem.",
+			value:           float64(dev.Size),
+			extraLabel:      []string{"device"},
+			extraLabelValue: []string{n},
+		})
+	}
+
+	// Information about data, metadata and system data
+	metrics = append(metrics, c.getAllocationStats("data", s.Allocation.Data)...)
+	metrics = append(metrics, c.getAllocationStats("metadata", s.Allocation.Metadata)...)
+	metrics = append(metrics, c.getAllocationStats("system", s.Allocation.System)...)
+
+	return metrics
+}
+
+// getAllocationStats returns allocation metrics for the given Btrfs Allocation statistics
+func (c *btrfsCollector) getAllocationStats(a string, s *btrfs.AllocationStats) []btrfsMetric {
+	metrics := []btrfsMetric{
+		{
+			name:            "reserved_bytes",
+			desc:            "Amount of space reserved for a data type",
+			value:           float64(s.ReservedBytes),
+			extraLabel:      []string{"type"},
+			extraLabelValue: []string{a},
+		},
+	}
+
+	// Add all layout statistics
+	for layout, stats := range s.Layouts {
+		metrics = append(metrics, c.getLayoutStats(a, layout, stats)...)
+	}
+
+	return metrics
+}
+
+// getLayoutStats returns metrics for a data layout
+func (c *btrfsCollector) getLayoutStats(a, l string, s *btrfs.LayoutUsage) []btrfsMetric {
+	return []btrfsMetric{
+		{
+			name:            "used_bytes",
+			desc:            "Amount of used space by a layout/data type",
+			value:           float64(s.UsedBytes),
+			extraLabel:      []string{"type", "mode"},
```

Comment:
We prefer to avoid `type` as a label name, it's too generic. What is this "type" exactly?

Reply:
"type" here is the "block group type", although I have also seen it referred to as "data type". I have renamed the label to `block_group_type`.